### DISCUSSION TITLE
TypeScript: Extracting keys from stylex.defineVars

### DISCUSSION THREAD

**Discussion by sumanbh** [facebook/stylex#466] [Q&A]
Is there a built-in type in stylex that helps us extract keys from stylex.defineVars?

Given this example:
```tsx
const test = stylex.defineVars({
  primary: 'red',
  secondary: 'blue',
});
```

Using TS's `keyof typeof x` results in this type (not ideal):
```tsx
type TestToken = keyof typeof test;

// type TestToken = "__opaqueId" | "__tokens" | typeof Symbol.toPrimitive | typeof Symbol.toStringTag | "primary" | "secondary" | "description" | "valueOf" | "toString"
```

We can manually exclude stylex's internal properties but I was wondering if there's a more elegant solution:

```tsx
type TestToken = Exclude<
  keyof typeof test,
  | '__opaqueId'
  | '__tokens'
  | 'description'
  | 'valueOf'
  | 'toString'
  | typeof Symbol.toPrimitive
  | typeof Symbol.toStringTag
>;

// type TestToken = "primary" | "secondary"
```

**Comment by nmn** [comment]
The `'__opaqueId'` and `'__tokens'` keys are an unfortunate requirement to make all the features work in TypeScript. Since TypeScript lacks support for `opaque types` this was needed to make this work.

A utility type to extract just the real keys from the object is easy to make, but I can't think of any use-cases where you would be able to use it. All usage of variables defined with `defineVars` has to be imported *by name*, directly from the module where they defined and then all keys have to be defined statically as well.

Any usage of the type `test[someVar]` are not allowed since the usages of the variables have to be known statically at compile time.

Please tell me your use-case in case you have something in mind that would work.

In any case, here's the utility type that would work:

```ts
type TActualKeys<T> = 
  T extends string ?
    T extends '__opaqueId' | '__tokens' ?
      never
    : T
  : never;

type VarGroupKeys<G extends VarGroup> = TActualKeys<keyof G>;

type TestToken = VarGroupKeys<typeof test>;
```

**Reply by neighborhood999** [reply]
Here is my solution:

<img width="860" alt="image" src="https://github.com/user-attachments/assets/9555c05f-cad4-4e7d-a80e-37f332d3b8ca">

```ts
import type * as stylex from '@stylexjs/stylex';

export type ExtractStylexVarKeys<T> =
  T extends stylex.VarGroup<infer K> ? keyof K : never;
```
